moshi/src/test/java/com/squareup/moshi/internal/RecursiveTypesResolveTest.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (C) 2017 Gson Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.squareup.moshi.internal;
+
+import com.squareup.moshi.JsonAdapter;
+import com.squareup.moshi.Moshi;
+import com.squareup.moshi.Types;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+/**
+ * Test fixes for infinite recursion on {@link Util#resolve(java.lang.reflect.Type, Class,
+ * java.lang.reflect.Type)}, described at <a href="https://github.com/google/gson/issues/440">Issue #440</a>
+ * and similar issues.
```

Comment:
I'm confused by this comment because the code that fixes it (in 7b73ce04ec6dcf329a42999a953f51db54ce3f92) puts the fixes in `Types`, not `Util`.

Reply:
Gson's `GsonTypes` has a bunch of stuff that's split into Moshi's `Types` and `Util`. In Moshi - this method is in `Util`